File: test-suite/src/test/java/io/micronaut/docs/jdbc/transactions/BookRepository.java
package io.micronaut.docs.jdbc.transactions;

// tag::imports[]
import jakarta.inject.Singleton;
import jakarta.transaction.Transactional;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
// end::imports[]

// tag::clazz[]
@Singleton
public class BookRepository {

    private final DataSource dataSource;

    public BookRepository(DataSource dataSource) { // <1>
        this.dataSource = dataSource;
    }

    @Transactional // <2>
    public void saveBook(Book book) throws SQLException {
        try (Connection connection = dataSource.getConnection(); // <3>
             PreparedStatement statement = connection.prepareStatement("INSERT INTO books (title, pages) VALUES (?, ?)")) {
            statement.setString(1, book.title());
            statement.setInt(2, book.pages());
            statement.executeUpdate();
        }
    }
    // end::clazz[]

    @Transactional
    public void createTable() throws SQLException {
        try (Connection connection = dataSource.getConnection();
             Statement statement = connection.createStatement()) {
            statement.execute("CREATE TABLE IF NOT EXISTS books (id BIGINT AUTO_INCREMENT PRIMARY KEY, title VARCHAR(255) NOT NULL, pages INT NOT NULL)");
        }
    }

    @Transactional
    public int count() throws SQLException {
        try (Connection connection = dataSource.getConnection();
             Statement statement = connection.createStatement();
             ResultSet resultSet = statement.executeQuery("SELECT COUNT(*) FROM books")) {
            resultSet.next();
            return resultSet.getInt(1);
        }
    }
    // tag::clazz[]
}
// end::clazz[]

File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/jdbc/transactions/BookRepository.kt
package io.micronaut.docs.jdbc.transactions

// tag::imports[]
import jakarta.inject.Singleton
import jakarta.transaction.Transactional
import javax.sql.DataSource
// end::imports[]

// tag::clazz[]
@Singleton
open class BookRepository(private val dataSource: DataSource) { // <1>

    @Transactional // <2>
    open fun saveBook(book: Book) {
        dataSource.connection.use { connection -> // <3>
            connection.prepareStatement("INSERT INTO books (title, pages) VALUES (?, ?)").use { statement ->
                statement.setString(1, book.title)
                statement.setInt(2, book.pages)
                statement.executeUpdate()
            }
        }
    }
    // end::clazz[]

    @Transactional
    open fun createTable() {
        dataSource.connection.use { connection ->
            connection.createStatement().use { statement ->
                statement.execute("CREATE TABLE IF NOT EXISTS books (id BIGINT AUTO_INCREMENT PRIMARY KEY, title VARCHAR(255) NOT NULL, pages INT NOT NULL)")
            }
        }
    }

    @Transactional
    open fun count(): Int {
        dataSource.connection.use { connection ->
            connection.createStatement().use { statement ->
                statement.executeQuery("SELECT COUNT(*) FROM books").use { resultSet ->
                    resultSet.next()
                    return resultSet.getInt(1)
                }
            }
        }
    }
    // tag::clazz[]
}
// end::clazz[]
